Breaking tokens into pieces should behave similar to Parser::bump.

tmiasko opened the pull request:
Previously when breaking tokens into smaller pieces, the replace_token
function have been used. It replaced current token and updated span
information, but it did not clear the list of expected tokens, neither
did it update remaining info about last token. This could lead to
incorrect error message, like one described in the issue #24780:

    expected one of ... `>` ...  found `>`

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented:
Thanks! Can you add a test case as well?

nagisa commented on src/libsyntax/parse/parser.rs:
```diff
@@ -977,15 +977,23 @@ impl<'a> Parser<'a> {
+        // for tokens currently using `bump_with`, last_token will be of no
+        // use anyway.
+        self.last_token = None;
+        self.last_token_interpolated = self.token.is_interpolated();
```
If you’re setting `last_token` to `None`, it makes little sense to store current token’s interpolation, right? This probably should be false unconditionally.

tmiasko commented:
I added test based on #24780 and changed last_token_interpolated to be false
unconditionally. 

Initially I was reluctant to add a test for this, because it seems to be
potentially fragile to changes in a parser (i.e., order of expected tokens), so
let me know if you have an idea for a better test.

alexcrichton commented on src/test/parse-fail/issue-24780.rs:
```diff
@@ -0,0 +1,20 @@
+// expected one of ..., `>`, ... found `>`
+//
+// compile-flags: -Z parse-only
+// error-pattern: expected one of `!`, `::`, `where`, or `{`, found `>`
```
We typically use `//~ ERROR` instead of `error-pattern` as it matches exhaustively and also allows checking span (line number) information

tmiasko commented:
Now using `//~^ ERROR` instead of `error-pattern`.

alexcrichton commented:
@bors: r+ cecf83f5926898c86c305c7aff104fcc900a87a2

bors commented:
:hourglass: Testing commit cecf83f with merge 5d771cd...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8021), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8042), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1289), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7982), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7992), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7982), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1195), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3147), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8027), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8030), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7995), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/165), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3035), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3051), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3042), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3024), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2257), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3029)
